### PR TITLE
Javascript only UI build cleanup

### PR DESCRIPTION
**Pull request by jackgr**
This PR contains only the javascript portions of #8776. It is part of the UI cleanup requested by #7122.

It does the following:
- Cleans up the npm/gulp based javascript build , which had a number of issues.
- Creates a scriptable version of the javascript build that doesn't run live reload and wait for changes.
- Separates build output from source artifacts, which were mixed in the original.

All of the changes are confined to the www directory. It doesn't change any of the build scripts in build/ or hack/, or affect the go build process, tools or outputs in any way.

### CODE REVIEW

File: www/master/gulpfile.js [hunk 1]
```diff
@@ -227,14 +262,14 @@ gulp.task('scripts:app:base', function() {
       .pipe(concat(build.scripts.app.main))
       .pipe(ngAnnotate())
       .on("error", handleError)
-      .pipe(isProduction ? uglify({preserveComments: 'some'}) : gutil.noop())
+      // .pipe(isProduction ? uglify({preserveComments: 'some'}) : gutil.noop())
```
Comment (bgrant0607) [inline www/master/gulpfile.js]: Why?

File: www/master/gulpfile.js [hunk 2]
```diff
@@ -342,20 +376,55 @@ gulp.task('copy:components', function() {
...
+
+gulp.task('content:html:app', function() {
+  return gulp.src(source.html.app.source, {base: 'shared'})
+    .on("error", handleError)
+    .pipe(prettify({
+        indent_char: ' ',
+        indent_size: 4,
+        unformatted: ['a', 'sub', 'sup', 'b', 'i', 'u']
+    }))
+    // .pipe(htmlify({
```
Comment (bgrant0607) [inline www/master/gulpfile.js]: Again, why?

File: www/master/gulpfile.js [hunk 3]
```diff
@@ -342,20 +376,55 @@ gulp.task('copy:components', function() {
...
+
+gulp.task('content:html:views', function() {
+  return gulp.src(source.html.views.source, {base: 'shared'})
+    .on("error", handleError)
+    .pipe(prettify({
+        indent_char: ' ',
+        indent_size: 4,
+        unformatted: ['a', 'sub', 'sup', 'b', 'i', 'u']
+    }))
+    // .pipe(htmlify({
```
Comment (bgrant0607) [inline www/master/gulpfile.js]: Also here

File: www/master/gulpfile.js (hunk 2 above)
Comment (jackgr) [inline www/master/gulpfile.js]: Explain in the code why this source is commented out.

**Comment by jackgr** [comment]
@bgrant0607 Removed/added comments as discussed and rebased.

**Comment by bgrant0607** [comment]
Thanks much.
LGTM.

**Comment by bcbroussard** [comment]
A couple minor things:
- Since `development.json` is .gitignored, shouldn't `www/master/shared/config/production.example.json` be `www/master/shared/config/development.example.json`? And shouldn't it look similar to `production.json`?
- Could you update the README with how to properly use development.json and production.json?

**Comment by bgrant0607** [comment]
Thanks for the review @bcbroussard

**Comment by rjnagal** [comment]
we'll wait for fixes for @bcbroussard comments before merging.

**Comment by jackgr** [comment]
@bcbroussard: Agreed. Done.

**Comment by jackgr** [comment]
Added the versions of the generated files created by the new build, since they are different due to pretty printing, uglification, minification, etc., so that running the ui build on a fresh clone of the repository yields no changes.

File: www/README.md [hunk 4]
```diff
@@ -36,25 +39,52 @@ For development you can serve the files locally by installing a webserver as fol
...
+
+```
+npm run build
+```
+
+Like `npm start`, it runs `bower install` to install and/or update the framework dependencies, but then it runs `gulp build` to generate a production version of the application. The `build` target builds the application for production (e.g., with uglification of js files and minification of css files), and does not run a file watcher, so that it can be used in automated build environments.
+
+To make the production code available to the Kubernetes api server, run this command from the top level directory:
+
+```hack/build-ui.sh```
```
Comment (bcbroussard) [inline www/README.md]: Markdown formatting is broken for the rest of the document. There needs to be a line break on both ends of the ticks.

File: www/master/components/README.md [hunk 5]
```diff
@@ -124,5 +130,4 @@ Sub-directories of this path are watched for changes, but not directly imported.
 Content available under the [CC-By 3.0
 license](http://creativecommons.org/licenses/by/3.0/)
```
Comment (bcbroussard) [inline www/master/components/README.md]: Can you remove the CC license? I don't think that is correct.

File: www/README.md (hunk 4 above)
Comment (jackgr) [inline www/README.md]: Thanks.

File: www/master/components/README.md (hunk 5 above)
Comment (jackgr) [inline www/master/components/README.md]: Done.

**Comment by bgrant0607** [comment]
Anything else @bcbroussard ?

**Comment by bcbroussard** [comment]
I spoke with @jackgr offline about adding a couple of small changes:
- building the development version should fail if a local `development.json` is missing.
- updating the gulp task to run the file copy tasks synchronously instead of the default async, which fixes an occasional race condition.

Those changes have been pushed in the latest commit. Thanks for the updates @jackgr.

LGTM :+1:

**Comment by jackgr** [comment]
Rebased.

**Comment by bgrant0607** [comment]
Thanks. LGTM, assuming tests pass.

**Comment by preillyme** [comment]
LGTM

**Pull request merged**